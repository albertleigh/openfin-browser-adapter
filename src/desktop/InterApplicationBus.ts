import {BaseApiClass} from './Base';

declare const Reflect:any;

export class InterApplicationBus extends BaseApiClass{
    static staticMethods:string[]=[
        'addUnsubscribeListener',
        'removeSubscribeListener',
        'removeUnsubscribeListener',
        'send',
        'unsubscribe',
    ];

    static instanceMethods:string[]=[];

    static subscribeListeners:Function[]=[];

    static listeners:{[key:string]:Function[]}={};

    static connection:any=null;

    static publish(topic:string, message:string, callback:Function){
        const {connection} = InterApplicationBus;

        if (!connection){
            console.log('no connection');
            return;
        }

        if (!Reflect.has(InterApplicationBus,'connection')){
            setTimeout(()=>{
                InterApplicationBus.publish(topic, message, callback);
            },10);
        }
    }

    static addSubscriberListener(listener:Function){
        if (!InterApplicationBus.connection){
            console.log('no conneciton');
            return;
        }

        InterApplicationBus.subscribeListeners.push(listener);
    }

    static subscribe(...args:any[]){
        if (!InterApplicationBus.connection){
            console.log('no connection');
            return;
        }

        let topic:string;
        let listener;

        if(args.length===4){
            [,,topic,listener] = args;
        }else{
            [,topic,listener] = args;
        }

        if (!Reflect.has(InterApplicationBus.listeners,topic)){
            InterApplicationBus.listeners[topic]=[];
        }

        InterApplicationBus.listeners[topic].push(listener);

        InterApplicationBus.subscribeListeners.forEach(lstnr=>(
           lstnr(null,topic,null)
        ));

    }

}