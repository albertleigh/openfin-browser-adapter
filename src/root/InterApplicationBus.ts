import {BaseApiClass} from './Base';
import {Channel} from "./Channel";
import {OpenFinMockV2} from '../OpenFinMockV2'

declare const Reflect:any;

type SubscribeListener = (uuid:string,topic:string,name:string)=>void;
type TopicSubscribeListener = (message:string,uuid:string,name:string)=>void;

export class InterApplicationBus extends BaseApiClass{
    static staticMethods:string[]=[
        'addUnsubscribeListener',
        'removeSubscribeListener',
        'removeUnsubscribeListener',
        'send',
        'unsubscribe',
    ];

    static instanceMethods:string[]=[];

    static subscribeListeners:SubscribeListener[]=[];

    static listeners:{[key:string]:TopicSubscribeListener[]}={};

    // static connection:any=null;

    static Channel = OpenFinMockV2.generateMethods('InterApplicationBus.Channel',Channel);

    static async publish(topic:string, message:string){

        if(InterApplicationBus.listeners[topic]){
            InterApplicationBus.listeners[topic].forEach((topicCb:TopicSubscribeListener)=>{
                topicCb(message,'','')
            })
        }
        InterApplicationBus.subscribeListeners.forEach((cb:SubscribeListener)=>{
            cb('',topic,'')
        });

        return;

    }

    static async addSubscribeListener(listener:SubscribeListener){
        InterApplicationBus.subscribeListeners.push(listener);
        return;
    }

    static async subscribe(
        senderUuid:string,name:string,topic:string,
        listener:TopicSubscribeListener
    ){

        if (!InterApplicationBus.listeners[topic]){
            InterApplicationBus.listeners[topic]=[];
        }

        InterApplicationBus.listeners[topic].push(listener);
        return;
    }

}