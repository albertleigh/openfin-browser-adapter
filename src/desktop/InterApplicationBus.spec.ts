import { InterApplicationBus } from './InterApplicationBus';

describe('Desktop::InterApplicationBus',()=>{

    let clazz:typeof InterApplicationBus;

    beforeEach(()=>{
        clazz = InterApplicationBus;
    });

    it('all unsupported static methods defined',()=>{
        expect(clazz.staticMethods).toMatchSnapshot();
    })

    it('all unsupported instance methods defined',()=>{
        expect(clazz.instanceMethods).toMatchSnapshot();
    })

});