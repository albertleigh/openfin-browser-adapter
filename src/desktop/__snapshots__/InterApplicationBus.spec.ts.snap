// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Desktop::InterApplicationBus all unsupported instance methods defined 1`] = `Array []`;

exports[`Desktop::InterApplicationBus all unsupported static methods defined 1`] = `
Array [
  "addUnsubscribeListener",
  "removeSubscribeListener",
  "removeUnsubscribeListener",
  "send",
  "unsubscribe",
]
`;
